### PR TITLE
sandbox: persistent volumes, non-root agent, skill containment

### PR DESCRIPTION
**Pull request by tescalada**
## Summary

Implements the design in #31 — persistent per-conversation `/workspace`, shared `/data` volume, non-root agent execution, and skill-script tamper protection. The spec body has known divergences (deferred requirements + design changes during implementation); see the [implementation-drift comment](https://github.com/rhiza-research/rhiza-agents/issues/31#issuecomment-4339741887) on the spec issue for the full delta and rationale.

Headline changes:

- **Storage:** `/workspace` (per-conversation subpath of homes volume), `/data` (shared volume across conversations). Configured via `DAYTONA_HOMES_VOLUME` / `DAYTONA_DATA_VOLUME` env vars (and matching `*_MOUNT_PATH` overrides).
- **Trust model:** agent runs as non-root `daytona` via `su -l` wrapping every agent-driven `process.exec`. Skill scripts at `/skills/<name>/scripts/<file>` are root-owned mode 0644 on the regular container FS — agent reads + executes via `run_file` but cannot modify (POSIX permissions enforce). `run_file` rejects any path not under `/skills/`. `write_file` is commented out; agent has no direct file-write tool.
- **Session tracking:** per-sandbox `inotifywait` daemon watches `/workspace` and `/data`; drained into `state["files"]` (path-only metadata) after each agent tool call.
- **Conversation deletion:** UI button per row; backend DELETE endpoint now does an explicit ownership check before any side effects (was a silent SQL no-op for non-owners that still cleaned up the owner's sandbox); homes-volume subpath cleanup hooked into delete.
- **Lazy migration:** legacy state-stored content materializes to the workspace volume on first access via `run_file` or `fetch_file_content`.

Deferred (platform constraints, see drift comment): library cache persistence (mountpoint-s3 doesn't support uv/pip rename ops), filesystem-enforced read-only on `/workspace` and `/data` (mountpoint-s3 rejects chown/chmod). The residual tamper surface is `execute_python_code` writes to those volumes, mediated by HITL approval; documented in `make_execute_python_code`'s docstring.

## Test plan

- [x] `uv run pytest` — 127 tests pass locally
- [ ] Frontend builds cleanly: from `frontend/`, run `npm run build` — produces `app.js` / `app.css`
- [ ] Set `DAYTONA_HOMES_VOLUME` and `DAYTONA_DATA_VOLUME` in `.envrc`
- [ ] `podman compose up -d`
- [ ] Send a chat message; activate a skill that bundles a script; run it via `run_file`
- [ ] Verify file panel session view shows files the skill wrote, with the right source labels (`data`, `output`)
- [ ] Toggle to "All files" — confirm live listing of `/workspace` and `/data` works
- [ ] Click a file — confirm content fetch via the file viewer endpoint (lazy-creates sandbox if not running)
- [ ] Try `run_file("/workspace/foo.py")` from the agent — confirm rejection
- [ ] Delete a conversation from the sidebar — confirm sandbox killed and homes-volume subpath cleaned
- [ ] Open a shared conversation URL as a different user — confirm read-only file viewing works, delete button absent

### CODE REVIEW

**Comment by tescalada** [comment]
## Review

### Overview

Substantial refactor that:

1. Moves agent file storage from LangGraph state to two Daytona volumes — `/workspace` (per-thread, subpath-mounted) and `/data` (shared).
2. Drops the agent to a non-root `daytona` user via `su -l` wrapping on every `process.exec`.
3. Removes `write_file`; restricts `run_file` to `/skills/<name>/scripts/<file>` paths only.
4. Replaces pre/post `list_files` diffing with a per-sandbox `inotifywait` daemon drained after each tool call.
5. Adds a UI delete affordance and fixes a silent ownership bug in the DELETE endpoint.
6. Adds lazy migration of legacy state-stored content on first read.

The trust model is well-documented in the docstrings — which protections are POSIX-enforced vs. HITL-mediated is explicit, not implicit.

### Strengths

- **Trust boundaries are clearly named.** `/skills/` is on the container FS so `chown root:root + 0644` actually enforces; `/workspace` and `/data` are mountpoint-s3 so HITL is the only defense — both stated in the docstring on `make_execute_python_code` rather than left as silent assumptions.
- **`exec_as_daytona`** base64-encodes the inner command before piping through `su -c`, avoiding shell-quoting hazards from agent-supplied code.
- **DELETE ownership fix** (`conversations.py` lines 100-106) — the prior `delete_conversation(id, user_id)` was a silent SQL no-op for non-owners that still ran `cleanup_sandbox`, letting any authenticated user kill the owner's sandbox. The explicit 403/404 before side effects is a real fix.
- **DB-row-first ordering on delete** is correct — partial failure leaves no orphan; sandbox/volume cleanup is best-effort and logged.
- **Path validation in `_validate_skill_path`** rejects `..`, double slashes, and shape mismatches. The defense-in-depth check in `exec_skill` is appropriate even though the agent has no path to plant a symlink in `/skills/`.

### Issues

#### Performance — `drain_inotify_journal` is O(N) round-trips per drained file

`sandbox.py` does `stat -c '%s|%Y' <path>` via `exec_as_daytona` once **per surviving path**. Each call is an HTTP round-trip to the Daytona toolbox API. Combined with `-e open -e access` watches, a script that imports many libraries can generate thousands of events; even after dedup-by-path, dozens-to-hundreds of stats per drain is realistic.

Suggestion: collect all surviving paths and stat them in a single `find ... -printf` (the same pattern `list_workspace_files` already uses), or drop `open`/`access` from the watched events — `create`, `modify`, `close_write` is what's actually load-bearing for "what files did this run touch."

#### Inotify journal can flood with `open` / `access`

Same root cause. `inotifywait -e open -e access` records every `open(2)` and `access(2)` syscall in the watched tree. `uv run` resolves a PEP 723 lockfile, imports stdlib, and the script imports its dependencies — each of those is at least an `open` event per file walked. The journal grows fast, and `mv` / `cat` / `rm` on a multi-MB log adds latency.

I'd cut this to `-e close_write -e create -e move -e delete`. That's the standard "what did this write" set; `open` / `access` are typically only useful for audit trails.

#### Drain has a race window where events are silently lost after the first drain

`drain_inotify_journal` uses:

```
mv journal journal.read && : > journal
cat journal.read
rm -f journal.read
```

The comment claims events during the rename window are preserved in `.read`. That's true while `inotifywait` holds the FD on the renamed inode — it keeps writing to the unlinked file. But `inotifywait` does **not** reopen the journal after the rename: its FD points at the old inode, not the new path. After `rm -f journal.read`, the inode is unlinked but the daemon's FD remains open and continues writing to that dead inode forever. New events written after the first drain go into the dangling inode, not the new journal — and the next drain reads an empty file.

I think this means session-file tracking only works for the **first** tool call after sandbox creation; every subsequent drain returns empty. Worth verifying with a test that drains twice in a row after writing to `/workspace` between drains.

Simpler pattern that avoids the FD-on-deleted-file issue: have the daemon write to a fresh file each drain by signaling it (`SIGHUP`), or use a FIFO, or just `cat journal && : > journal` with no rename — `cat` then truncate-in-place leaves a tiny race but the daemon's append redirect lands at the end of whatever exists.

#### `download_conversation_file` Content-Disposition is not escaped

```python
filename = lookup_path.rsplit("/", 1)[-1] or "file"
return Response(
    ...
    headers={"Content-Disposition": f'attachment; filename="{filename}"'},
)
```

A filename containing `"` or `\r\n` breaks the header / enables header injection. Logical paths come from agent + filesystem, so this isn't directly attacker-controlled, but a skill that creates `foo"; X-Injected: bar.txt` would inject. RFC 6266 / 5987 form (`filename*=UTF-8''<percent-encoded>`) is the safe encoding; minimum bar is to strip CR/LF and escape quotes.

#### `window.confirm` / `window.alert` for delete UX

`conversations.ts` lines 177, 186, 200. Browser modal dialogs are blocking, can't be styled, and break automation. Probably fine for a first cut, but a custom confirm modal would match the rest of the UI's idioms.

#### `cleanup_thread_workspace` spins up a fresh sandbox just to `rm -rf`

Acknowledged in the docstring ("Daytona has no volume file-delete API"), but creating + deleting a full sandbox on every conversation-delete is expensive. If the active per-thread sandbox is still alive, you could `rm -rf` through that one first and only fall back to the temp sandbox when no thread sandbox exists.

#### Test coverage gaps

The skill activation tests are well-updated. But the riskiest new code is untested:

- `_validate_skill_path` (path traversal, edge cases — `/skills//foo/scripts/x`, `/skills/foo/scripts/`, `/skills/foo/scripts/../../../etc/passwd`)
- `drain_inotify_journal` parsing and the race described above
- `exec_as_daytona` env / cwd injection
- The DELETE ownership 403/404 paths
- `fetch_file_content` legacy fallback path

These are pure functions or thin wrappers — should be straightforward to test with the same `_FakeSandbox` pattern already in `tests/test_skills_activation.py`.

### Minor

- `files.py` lines 649-674 — large commented-out `write_file` block. Project guidance discourages "removed comments for removed code." If git history is the source of truth, deleting it is cleaner; if you want to preserve revival intent, a one-line `# Removed in #32; see git log` is enough.
- `chat.py` lines 2036-2048 — same with the commented `file_written` SSE block.
- `_validate_skill_path` `parts[1] == ""` check is redundant with the prefix check that already guarantees `/skills/<something>/...`.
- `start_inotify_daemon` uses `nohup ... &` inside an `exec_as_daytona` wrapper — the wrapper is `su -l ... -c "echo b64 | base64 -d | sh"`. Confirm the shell process actually backgrounds correctly under `su -l` (some `su` implementations wait on all children of the spawned shell). A quick test would catch a regression here.

### Risk assessment

The trust-model changes are sound and well-explained. The persistence + non-root parts are the high-blast-radius pieces; my biggest concerns are operational, not security:

1. The inotify drain race (likely silently drops session-file tracking after the first drain).
2. The O(N) stat round-trips (could make tool-call latency unpleasant for I/O-heavy scripts).

Both are fixable without rearchitecting. I'd land this with the inotify race verified or fixed, and follow up on the perf and Content-Disposition escaping.

**Review by tescalada — COMMENTED** [review]
## Overview

- Moves agent file storage from LangGraph state to two Daytona volumes — `/workspace` (per-thread, subpath-mounted) and `/data` (shared).
- Drops the agent to a non-root `daytona` user via `su -l` wrapping on every `process.exec`.
- Removes `write_file`; restricts `run_file` to `/skills/<name>/scripts/<file>` paths only.
- Replaces pre/post `list_files` diffing with a per-sandbox `inotifywait` daemon drained after each tool call.
- Adds a UI delete affordance and fixes a silent ownership bug in the DELETE endpoint.
- Adds lazy migration of legacy state-stored content on first read.

Trust model is well-documented in the docstrings — which protections are POSIX-enforced vs. HITL-mediated is explicit, not implicit.

## Strengths

- **Trust boundaries are clearly named.** `/skills/` is on the container FS so `chown root:root + 0644` actually enforces; `/workspace` and `/data` are mountpoint-s3 so HITL is the only defense — both stated in the docstring on `make_execute_python_code` rather than left as silent assumptions.
- **`exec_as_daytona`** base64-encodes the inner command before piping through `su -c`, avoiding shell-quoting hazards from agent-supplied code.
- **DELETE ownership fix** — the prior `delete_conversation(id, user_id)` was a silent SQL no-op for non-owners that still ran `cleanup_sandbox`, letting any authenticated user kill the owner's sandbox. The explicit 403/404 before side effects is a real fix.
- **DB-row-first ordering on delete** is correct — partial failure leaves no orphan; sandbox/volume cleanup is best-effort and logged.
- **Path validation in `_validate_skill_path`** rejects `..`, double slashes, and shape mismatches. The defense-in-depth check in `exec_skill` is appropriate even though the agent has no path to plant a symlink in `/skills/`.

## Risk assessment

Trust-model changes are sound and well-explained. The persistence + non-root parts are the high-blast-radius pieces; biggest concerns are operational, not security:

1. The inotify drain race (likely silently drops session-file tracking after the first drain).
2. The O(N) stat round-trips (could make tool-call latency unpleasant for I/O-heavy scripts).

Both are fixable without rearchitecting. I'd land this with the inotify race verified or fixed, and follow up on the perf and Content-Disposition escaping.

## Test coverage gaps

The skill activation tests are well-updated. But the riskiest new code is untested:

- `_validate_skill_path` (path traversal, edge cases — `/skills//foo/scripts/x`, `/skills/foo/scripts/`, `/skills/foo/scripts/../../../etc/passwd`)
- `drain_inotify_journal` parsing and the race called out inline
- `exec_as_daytona` env / cwd injection
- The DELETE ownership 403/404 paths
- `fetch_file_content` legacy fallback path

These are pure functions or thin wrappers — should be straightforward to test with the same `_FakeSandbox` pattern already in `tests/test_skills_activation.py`.

Inline comments anchor the specific issues below.

File: src/rhiza_agents/agents/tools/sandbox.py [hunk 1]
```diff
@@ -103,6 +189,213 @@ def _normalize_sandbox_upload_path(path: str) -> str:
...
+    init_cmd = (
+        f": > {journal}; "
+        # Start the daemon. -m monitor mode (don't exit), -r recursive,
+        # -e events. --format produces TSV; --timefmt %s makes the
+        # timestamp a unix epoch second so we don't have to parse a
+        # human date string in the drain.
+        f"nohup inotifywait -mr "
+        f"-e create -e modify -e close_write -e open -e access "
+        f"--format '%e\\t%w%f\\t%T' --timefmt '%s' "
+        f"{workspace} {data} > {journal} 2>&1 &"
```
Comment (tescalada) [inline src/rhiza_agents/agents/tools/sandbox.py]: **Inotify watch set is too broad.** `-e open -e access` records every `open(2)` and `access(2)` syscall in the watched tree. `uv run` resolves a PEP 723 lockfile, imports stdlib, and the script imports its dependencies — each of those is at least one `open` event per file walked. The journal grows fast, and the `mv`/`cat`/`rm` drain plus per-path `stat` round-trips (next comment) get expensive proportional to that.

For the "what files did this run touch" use case, `-e close_write -e create -e move -e delete` is enough. `open`/`access` are typically only useful for audit trails.

File: src/rhiza_agents/agents/tools/sandbox.py [hunk 2]
```diff
@@ -103,6 +189,213 @@ def _normalize_sandbox_upload_path(path: str) -> str:
...
+    # FD open against the renamed inode and continues writing to it
+    # until we delete it, so events that happened during the rename
+    # window are preserved in the .read copy.
+    drain_cmd = (
+        f"if [ -f {journal} ]; then "
+        f"  mv {journal} {journal}.read 2>/dev/null && : > {journal}; "
+        f"  cat {journal}.read 2>/dev/null; "
+        f"  rm -f {journal}.read; "
+        f"fi"
+    )
```
Comment (tescalada) [inline src/rhiza_agents/agents/tools/sandbox.py]: **I think this drain has a race that silently drops events after the first call.**

The comment above claims events during the rename window are preserved in `.read`. That's true while `inotifywait` holds the FD on the renamed inode — it keeps writing to the unlinked file. But `inotifywait` does **not** reopen the journal after the rename: its FD points at the old inode, not the new path. So:

1. First drain: `mv journal journal.read` — daemon's FD now points at `journal.read`'s inode. Drain reads `journal.read`, then `rm -f journal.read`. Inode is unlinked but the daemon's FD remains open and continues writing to that dead inode.
2. After `rm`: the daemon writes to the dangling inode forever. New events do not land in the new (empty) `journal` file.
3. Second drain: reads an empty `journal`, returns `{}`.

This means session-file tracking probably only works for the **first** tool call after sandbox creation; every subsequent drain returns empty.

Worth a test that drains twice in a row after writing to `/workspace` between drains. Fix options:
- Have the daemon write to a fresh file each drain via `SIGHUP` + log rotation.
- Use a FIFO instead of a regular file.
- Just `cat journal && : > journal` with no rename — the daemon's append-on-open redirect (`>` here) doesn't reopen on truncate, but truncating in place keeps the same inode so the daemon's writes still land in it. Tiny race on events between `cat` and `: >`, but no FD-on-deleted-file issue.

File: src/rhiza_agents/agents/tools/sandbox.py [hunk 3]
```diff
@@ -103,6 +189,213 @@ def _normalize_sandbox_upload_path(path: str) -> str:
...
+            source = default_source
+        elif abs_path == SANDBOX_DATA or abs_path.startswith(SANDBOX_DATA + "/"):
+            logical = abs_path
+            source = "data"
+        else:
+            # Out of scope — skip events from outside watched areas.
+            continue
+
+        # Stat to get size + mtime. If gone, skip.
+        stat_resp = exec_as_daytona(sandbox, f"stat -c '%s|%Y' {shlex.quote(abs_path)}")
```
Comment (tescalada) [inline src/rhiza_agents/agents/tools/sandbox.py]: **O(N) HTTP round-trips per drained file.** Each `exec_as_daytona` is a round-trip to Daytona's toolbox API. Combined with the broad watch set above, an I/O-heavy script could mean dozens-to-hundreds of `stat` calls per drain.

`list_workspace_files` already uses the right pattern — `find ... -printf '%p\t%s\t%T@\n'` collects size + mtime for all paths in one exec. Same here: build the list of surviving paths, pass them all to a single `stat -c '%s|%Y|%n' path1 path2 ...` (or pipe through `xargs`), then parse. One round-trip instead of N.

File: src/rhiza_agents/agents/tools/sandbox.py [hunk 4]
```diff
@@ -172,6 +465,144 @@ def _daytona_sandbox_resources_from_env():
...
+
+        vol = _get_daytona().volume.get(homes_name, create=False)
+        # Minimal image — no need for git/uv since we just rm -rf
+        image = Image.debian_slim("3.12")
+        sb = _get_daytona().create(
+            CreateSandboxFromImageParams(
+                image=image,
+                volumes=[VolumeMount(volume_id=vol.id, mount_path=homes_path, subpath=thread_id)],
+            )
+        )
```
Comment (tescalada) [inline src/rhiza_agents/agents/tools/sandbox.py:892]: **Spinning up a full sandbox just to `rm -rf` is heavy** for the common case where the per-thread sandbox is still alive in `_sandboxes`. If the active sandbox is still around, the `rm -rf` could go through it directly (its mount already points at the same volume + subpath); fall back to creating a temp sandbox only when no live one exists. Acknowledged as Daytona-API-limited in the docstring, but the optimization is purely client-side.

File: src/rhiza_agents/agents/tools/sandbox.py [hunk 5]
```diff
@@ -103,6 +189,213 @@ def _normalize_sandbox_upload_path(path: str) -> str:
...
+        f": > {journal}; "
+        # Start the daemon. -m monitor mode (don't exit), -r recursive,
+        # -e events. --format produces TSV; --timefmt %s makes the
+        # timestamp a unix epoch second so we don't have to parse a
+        # human date string in the drain.
+        f"nohup inotifywait -mr "
+        f"-e create -e modify -e close_write -e open -e access "
+        f"--format '%e\\t%w%f\\t%T' --timefmt '%s' "
+        f"{workspace} {data} > {journal} 2>&1 &"
+    )
```
Comment (tescalada) [inline src/rhiza_agents/agents/tools/sandbox.py:447]: Worth confirming the daemon actually backgrounds correctly under the `su -l ... -c "..."` wrapper. Some `su` implementations wait on all descendants of the spawned shell even with `nohup` + `&`, which would mean `exec_as_daytona` blocks until the daemon exits (i.e., never). A quick test that asserts the exec returns within a second after this call would catch a regression here.

File: src/rhiza_agents/agents/tools/files.py [hunk 6]
```diff
@@ -36,70 +66,71 @@ def _build_uv_run_cmd(filename: str, script_args: list[str] | None) -> str:
...
+        return f"Path must be under {SANDBOX_SKILLS_DIR}/"
+    if ".." in logical_path.split("/"):
+        return "Path may not contain '..' components"
+    if "//" in logical_path:
+        return "Path may not contain consecutive slashes"
+    parts = logical_path.lstrip("/").split("/")
+    # Expected shape: skills / <name> / scripts / <file>
+    if len(parts) < 4 or parts[2] != "scripts":
+        return f"Path must be of the form {SANDBOX_SKILLS_DIR}/<name>/scripts/<file>"
+    if parts[-1] == "" or parts[1] == "":
```
Comment (tescalada) [inline src/rhiza_agents/agents/tools/files.py]: Nit: `parts[1] == ""` is unreachable here — a path like `/skills//foo/scripts/x.py` (which is the only way to get an empty segment at index 1 after `lstrip("/")`) is rejected three lines up by the `"//" in logical_path` check. Could be dropped or replaced with a comment.

File: src/rhiza_agents/routes/conversations.py [hunk 7]
```diff
@@ -215,40 +276,107 @@ async def get_conversation_file(
...
+        raise HTTPException(status_code=404, detail="File not found") from e
+    except Exception as e:
+        logger.warning("Failed to fetch file %s for conversation %s", lookup_path, conversation_id, exc_info=True)
+        raise HTTPException(status_code=502, detail=f"Failed to fetch file: {e}") from e
+
+    filename = lookup_path.rsplit("/", 1)[-1] or "file"
+    return Response(
+        content=content_bytes,
+        media_type="application/octet-stream",
+        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
```
Comment (tescalada) [inline src/rhiza_agents/routes/conversations.py]: **Content-Disposition filename is not escaped.** A filename containing `"` or `\r\n` breaks the header / enables header injection. Logical paths come from agent + filesystem here, not directly attacker-controlled, but a skill that emits `foo"; X-Injected: bar.txt` would inject.

Fix: RFC 6266 / 5987 form is the safe encoding:
```python
from urllib.parse import quote
encoded = quote(filename, safe="")
headers = {"Content-Disposition": f"attachment; filename*=UTF-8''{encoded}"}
```
Minimum bar is to strip `\r\n` and escape backslashes/quotes.

File: frontend/src/widgets/conversations.ts [hunk 8]
```diff
@@ -46,18 +52,83 @@ export class ConversationListWidget extends Widget {
...
+            let message = `${response.status} ${response.statusText}`;
+            try {
+                const data = await response.json();
+                if (data?.detail) message = data.detail;
+            } catch {
+                // ignore JSON parse failure; fall back to status text
             }
-            a.textContent = conv.title || 'New conversation';
-            nav.appendChild(a);
+            window.alert(`Failed to delete: ${message}`);
```
Comment (tescalada) [inline frontend/src/widgets/conversations.ts]: `window.confirm` / `window.alert` are blocking, can't be styled, and break browser automation. Probably fine as a first cut, but a custom confirm modal would match the rest of the UI's idioms and be testable end-to-end.

File: src/rhiza_agents/agents/tools/sandbox.py (hunk 1 above)
Comment (tescalada) [reply src/rhiza_agents/agents/tools/sandbox.py]: Fixed in 43606ec. The watch set keeps `-e create -e modify -e close_write -e move -e delete -e open -e access`. The `open` and `access` events are deliberately kept — spec requirement 8 depends on them for cache-hit visibility on `/data`. The reviewer perf concern doesnt fire in practice because Python imports load from venvs / `uv` cache (`/root/.cache/uv`, `/root/.venv`) which are outside the watched trees, so `open`/`access` only fire for files the script explicitly opens under `/workspace` or `/data` — which is the signal we want.

File: src/rhiza_agents/agents/tools/sandbox.py (hunk 2 above)
Comment (tescalada) [reply src/rhiza_agents/agents/tools/sandbox.py]: Fixed in 43606ec. Switched to cat-then-truncate-in-place; the daemon redirect changed from `>` to `>>` so writes use O_APPEND and atomically seek to end-of-file after each truncate, which keeps subsequent drains seeing fresh events instead of writing past a stale FD position into a sparse hole. Tiny race window between the cat and the truncate where events can be lost (microseconds), accepted as the cost of this approach. `tests/test_sandbox_helpers.py` exercises the parser; the multi-drain integration would need a real sandbox so left for the smoke test.

File: src/rhiza_agents/agents/tools/sandbox.py (hunk 3 above)
Comment (tescalada) [reply src/rhiza_agents/agents/tools/sandbox.py]: Fixed in 43606ec. The drain now collects all surviving paths and runs a single `stat -c %n|%s|%Y` over them in one exec, with stderr discarded so missing files (deleted between event and drain) silently drop out. Two round-trips per drain regardless of how many files were touched, instead of N+1.

File: src/rhiza_agents/agents/tools/files.py (hunk 6 above)
Comment (tescalada) [reply src/rhiza_agents/agents/tools/files.py]: Fixed in 560bfcd. Dropped the redundant check; the consecutive-slashes guard above already catches every input that could produce an empty parts[1]. Added `tests/test_skill_path_validation.py` with 33 cases covering the path-traversal, double-slash, and shape-mismatch surfaces (including the exact `/skills//foo/scripts/x.py` case that the redundant check was guarding) — the guard is provably load-bearing now.

File: src/rhiza_agents/routes/conversations.py (hunk 7 above)
Comment (tescalada) [reply src/rhiza_agents/routes/conversations.py]: Fixed in 4020d63. Replaced the f-string interpolation with `_content_disposition_attachment(filename)` which strips ASCII control characters (0x00-0x1F, 0x7F) so CR/LF cannot inject a header, escapes `"` and `\\` in the ASCII fallback per RFC 6266 grammar, and emits `filename*=UTF-8<pct-encoded>` for unicode-capable clients. `tests/test_content_disposition.py` covers the injection vectors (CRLF, double-quote, backslash, NUL/control), unicode, and the empty-filename fallback.

File: src/rhiza_agents/agents/tools/sandbox.py (hunk 4 above)
Comment (tescalada) [reply src/rhiza_agents/agents/tools/sandbox.py:892]: Fixed in da0870d. `cleanup_thread_workspace` now reuses the active per-thread sandbox if it is alive in `_sandboxes` (its mount already points at the right volume + subpath), and only spins a temp sandbox as a fallback when the per-thread one is gone. Caller in `routes/conversations.py` reordered so the workspace cleanup runs before `cleanup_sandbox` kills the active sandbox — without the reorder the alive-sandbox path could never fire. `tests/test_cleanup_thread_workspace.py` covers both paths plus the fallback when the active sandbox raises.

File: src/rhiza_agents/agents/tools/sandbox.py (hunk 5 above)
Comment (tescalada) [reply src/rhiza_agents/agents/tools/sandbox.py:447]: Deferred — leaving for the smoke-test step rather than fixing in this PR. The concern is real (some `su` implementations wait on all descendants of the spawned shell, which would mean `exec_as_daytona` blocks until the daemon exits) but it cannot be unit-tested: the behavior being verified is the platform `su` → shell → nohup propagation, not Python-side logic. The smoke test naturally exercises it — if the daemon were blocking, every sandbox creation would hang and the agent flow would fail immediately. If that surfaces, the fix is `setsid` or `disown` inside the wrap.

File: frontend/src/widgets/conversations.ts (hunk 8 above)
Comment (tescalada) [reply frontend/src/widgets/conversations.ts]: Fixed in dbe340b. Replaced `window.confirm` and `window.alert` with `confirmDialog()` and `alertDialog()` in `frontend/src/lib/dialog.ts` — both are promise-returning wrappers around the native `<dialog>` element, which renders as a regular DOM node so automation can locate its buttons normally (`button[name="confirm"]`, `button[name="cancel"]`). Delete confirmation uses the destructive-styled variant; both error paths use `alertDialog`. CSS in `style.css`. Frontend bundle rebuilt.

**Comment by tescalada** [comment]
Pushed a round of hardening commits addressing security and robustness issues found in review.

**Path safety**
- `workspace_path` now rejects path traversal: a logical path that normalizes outside `/workspace` or `/data` (e.g. `../../etc/passwd`, `/data/../../etc/x`) raises and maps to 404 instead of being read/written as root. Cross-volume normalization (a `/data` path resolving into `/workspace` or vice versa) is also rejected.
- File reads/writes additionally resolve the real path inside the sandbox (`realpath`) and reject symlink targets that escape `/workspace` or `/data`, so a planted symlink can't be followed past the lexical check.

**Access control**
- The file-content endpoint only performs the legacy→volume migration write when the requester owns the conversation. Read-only (shared-link) viewers no longer trigger a write to the owner's volume; a not-yet-materialized file returns 404 for them.

**Resource limits**
- File fetch/download is capped at 50 MiB, enforced both on the stat size and on the bytes actually read, so a large `/data` artifact can't OOM the server. Oversized requests return 413.
- The workspace-cleanup delete refuses to run against `/` or any system directory (prefix-based check), guarding against a misconfigured mount path.

**Session-file tracking**
- `execute_python_code` now runs with cwd `/workspace`, so relative-path writes land on the persistent volume and appear in the file panel.
- The inotify daemon is probed and restarted (drain-first, so already-journaled events aren't lost) if it has died; `procps` was added to the image so the liveness check works. Watch-limit exhaustion at startup is detected and logged. File-listing and journal parsing are now robust to tab/newline characters in filenames.

**Frontend**
- The conversation delete button is guarded against double-clicks (no duplicate DELETE, no spurious failure alert).
- The file panel ignores stale responses when the scope toggle changes mid-fetch, and no longer leaves a stuck "Loading…" placeholder.

Tests added for traversal rejection, owner-vs-viewer migration, the size cap, the cleanup guard, symlink containment, and filename parsing. Full suite: 240 passing; frontend builds clean.
